primitives/manta/src/currencies.rs

```diff
@@ -0,0 +1,185 @@
+where
+    C: Config,
+    A: AssetConfig<C>,
+    A::AssetId:
+        Parameter + Member + Copy + MaybeSerializeDeserialize + Ord + TypeInfo + MaxEncodedLen,
+    A::Balance: AtLeast32BitUnsigned
+        + FullCodec
+        + Copy
+        + MaybeSerializeDeserialize
+        + Debug
+        + Default
+        + scale_info::TypeInfo
+        + MaxEncodedLen,
+    Native: fungible::Inspect<C::AccountId, Balance = A::Balance>
+        + fungible::Mutate<C::AccountId>
+        + Currency<C::AccountId, Balance = A::Balance>,
+    NonNative: fungibles::Inspect<C::AccountId, AssetId = A::AssetId, Balance = A::Balance>
+        + Mutate<C::AccountId>
+        + Transfer<C::AccountId>,
+{
+    type CurrencyId = A::AssetId;
+    type Balance = A::Balance;
+
+    fn minimum_balance(currency_id: Self::CurrencyId) -> Self::Balance {
+        if currency_id == A::NativeAssetId::get() {
+            <Native as fungible::Inspect<C::AccountId>>::minimum_balance()
+        } else {
+            NonNative::minimum_balance(currency_id)
+        }
+    }
+
+    fn total_issuance(currency_id: Self::CurrencyId) -> Self::Balance {
+        if currency_id == A::NativeAssetId::get() {
+            <Native as fungible::Inspect<C::AccountId>>::total_issuance()
+        } else {
+            NonNative::total_issuance(currency_id)
+        }
+    }
+
+    fn total_balance(currency_id: Self::CurrencyId, who: &C::AccountId) -> Self::Balance {
+        if currency_id == A::NativeAssetId::get() {
+            Native::balance(who)
+        } else {
+            NonNative::balance(currency_id, who)
+        }
+    }
+
+    fn free_balance(currency_id: Self::CurrencyId, who: &C::AccountId) -> Self::Balance {
+        if currency_id == A::NativeAssetId::get() {
+            Native::free_balance(who)
+        } else {
+            NonNative::balance(currency_id, who)
+        }
+    }
+
+    fn ensure_can_withdraw(
+        currency_id: Self::CurrencyId,
+        who: &C::AccountId,
+        amount: Self::Balance,
+    ) -> DispatchResult {
+        if amount == Zero::zero() {
+            return Ok(());
+        }
+        let new_balance = Self::free_balance(currency_id, who)
+            .checked_sub(&amount)
+            .ok_or(DispatchError::Arithmetic(ArithmeticError::Underflow))?;
+        if currency_id == A::NativeAssetId::get() {
+            Native::ensure_can_withdraw(who, amount, WithdrawReasons::empty(), new_balance)
+        } else {
+            ensure!(
+                new_balance >= Self::minimum_balance(currency_id),
+                DispatchError::Other("balance too low")
+            );
+            Ok(())
+        }
+    }
+
+    fn transfer(
+        currency_id: Self::CurrencyId,
+        from: &C::AccountId,
+        to: &C::AccountId,
+        amount: Self::Balance,
+    ) -> DispatchResult {
+        if currency_id == A::NativeAssetId::get() {
+            Native::transfer(from, to, amount, ExistenceRequirement::KeepAlive)?;
```

Comment:
Maybe this should be AllowDeath ? Wouldn't it help with https://github.com/Manta-Network/Manta/issues/1172 actually

Reply:
Also on the withdraw, althought it doesn't seem to be used

Reply:
If we support staking native token, then Stake all operation will make account as
 dust account. I think this has issue that user can't execute furthure transaction as user don't have native token to paying fees.
I'll add a testcase for this case.